src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/domain/timeline/Timeline.kt

```diff
@@ -0,0 +1,34 @@
+package uk.gov.justice.digital.hmpps.educationandworkplanapi.domain.timeline
+
+/**
+ * A Timeline is sequenced list of [TimelineEvent]s for a given prisoner. The list can be empty, representing an empty
+ * timeline for the prisoner.
+ *
+ * The list of events cannot be mutated directly, and instead should be mutated through methods such as [addEvent]
+ */
+class Timeline(
+  val prisonNumber: String,
+  events: List<TimelineEvent> = emptyList(),
+) {
+
+  // _events is the real property and is deliberately private to prevent external access and mutation.
+  // events is a pseudo property that returns the current _events field as a List (ie. non-mutable)
+  private val _events: MutableList<TimelineEvent>
+  val events: List<TimelineEvent>
+    get() = _events.toList()
```

Comment:
I'm not entirely happy with this approach to prevent someone using the property directly and mutating it. Having the "real" property as private, and then a "pseudo" property seemed like an OK'ish approach (in the absence of finding anything native in the language to help); but I'm not entirely comfortable with the underscore prefix.

Happy for any comments or suggestions here

Reply:
If I've understood this correctly, you don't want to return the real List because that would allow someone to mutate it (a problem a naive developer often overlooks, because they think simply making the fields private is enough to make something immutable).

Another approach (which I think was recommended by Joshua Bloch, but it's a long time since I read his book!), is to return a copy of the list. Could you do that instead?

Reply:
> If I've understood this correctly, you don't want to return the real List because that would allow someone to mutate it
Yes, exactly. And the order/sequence of the list is the important thing which we want to maintain in the domain

But thinking about it, there might be a different approach - the key thing we're interested in is the order of the list of events, not necessarily preventing mutation. The idea of preventing mutation was to ensure we were in control of the list, and therefore the order. But I wonder if there's a different way to do this which would be easier to read and grok .... let me have a play ...

Reply:
I've just committed a change to this class to simplify the events list and the part around ensuring its an ordered list